app/models/base_facility.rb

```diff
@@ -35,6 +67,48 @@ def pull_source_data
       max_record_count = metadata['maxRecordCount']
       Facilities::SharedClient.new.get_all_facilities(*FACILITY_SORT_FIELDS[name], max_record_count).map(&method(:new))
     end
+
+    def find_facility_by_id(id)
+      type, unique_id = id.split('_')
+      return nil unless type && unique_id
+      "Facilities::#{type.upcase}Facility".constantize.find_by(unique_id: unique_id)
+    end
+
+    def query(params)
+      bbox_num = params[:bbox].map { |x| Float(x) }
+      data = build_result_set(bbox_num, params[:type], params[:services])
+      Common::Collection.new(::VAFacility, data: data.sort_by(&(dist_from_center bbox_num)))
+    end
+
+    def build_result_set(bbox_num, type, services)
+      lats = bbox_num.values_at(1, 3)
+      longs = bbox_num.values_at(2, 0)
+      conditions = { lat: coord_range(lats), long: coord_range(longs) }
+      TYPES.map { |facility_type| get_facility_data(conditions, type, facility_type, services) }.flatten
+    end
+
+    def get_facility_data(conditions, type, facility_type, services)
+      return [] unless type.blank? || type == facility_type
+      facilities = TYPE_MAP[facility_type].constantize.where(conditions)
+      facilities = facilities.where("services->'benefits'->'standard' @> '#{services}'") if services&.any?
+      facilities
+    end
```

Comment:
A FIXME comment or warning here would be helpful. When this returns early because `type.blank?` you're getting an empty Array object. But when it doesn't return early you get an ActiveRecord::Collection object, the latter is chainable whereas the former is not.

``` ruby
BaseFacility.get_facility_data.where(some_condition) # would fail when array
```

We should probably track this in a Github issue related to broader refactor of using PostGIS and not relying on `Common::Collection` to do things like `sort` and `filter.` Pushing as much off to the database as opposed to doing it in memory will be preferred. Though it's probably not going to be too big of an issue with 21000 records.

Reply:
I modified this to return <model>.none which is chainable :)